File: Data Structures & Algorithms/search-2d-matrix/submission-4.js
class Solution {
    /**
     * @param {number[][]} matrix
     * @param {number} target
     * @return {boolean}
     */
    searchMatrix(matrix, target) {
        let left = 0;
        let right = matrix.length - 1;
        let mid = 0;
        while(left <= right){
            mid = left + Math.floor((right - left) / 2);
            if(matrix[mid][0] > target){
                right = mid - 1
            }else if(matrix[mid][matrix[0].length - 1] < target){
                left = mid + 1
            }else{
                break
            }
        }
        console.log(mid)

        left = 0
        right = matrix[mid].length - 1;
        while(left <= right){
            let m = left + Math.floor((right - left) / 2);
            if(matrix[mid][m] < target){
                left = m + 1
            }else if(matrix[mid][m] > target){
                right = m - 1
            }else{
                return true;
            }
        }
        return false;
    }
}

File: Data Structures & Algorithms/find-target-in-rotated-sorted-array/submission-4.py
class Solution:
    def search(self, nums: List[int], target: int) -> int:
        left = 0
        right = len(nums) - 1
        min_val = float("inf")
        arr = nums
        while(left <= right):
            mid = (left + right) // 2
            if(target == nums[mid]):
                return mid
            elif(nums[left] <= nums[mid]):
                if(nums[left] > target or nums[mid] < target):
                    left = mid + 1
                else:
                    right = mid - 1
            else:
                if(nums[right] < target or nums[mid] > target):
                    right = mid - 1
                else:
                    left = mid + 1

        return -1